--- 2.6.24.3.base/drivers/md/dm.c	2008-03-19 23:48:32.000000000 -0700
+++ 2.6.24.3/drivers/md/dm.c	2008-03-20 00:49:48.000000000 -0700
@@ -889,6 +889,17 @@ static int dm_any_congested(void *conges
 	return r;
 }
 
+unsigned dm_inflight_total(struct dm_target *target)
+{
+	struct mapped_device *md;
+	request_queue_t *queue;
+	md = dm_table_get_md(target->table);
+	queue = md->queue;
+	dm_put(md);
+	return (queue->capacity - atomic_read(&queue->available));
+}
+EXPORT_SYMBOL_GPL(dm_inflight_total);
+
 /*-----------------------------------------------------------------
  * An IDR is used to keep track of allocated minor numbers.
  *---------------------------------------------------------------*/
--- 2.6.24.3.base/include/linux/device-mapper.h	2008-02-25 16:20:20.000000000 -0800
+++ 2.6.24.3/include/linux/device-mapper.h	2008-03-20 00:49:48.000000000 -0700
@@ -196,6 +196,7 @@ int dm_copy_name_and_uuid(struct mapped_
 struct gendisk *dm_disk(struct mapped_device *md);
 int dm_suspended(struct mapped_device *md);
 int dm_noflush_suspending(struct dm_target *ti);
+unsigned dm_inflight_total(struct dm_target *target);
 
 /*
  * Geometry functions.
