--- /tmp/tmp.fzMagE5020/a/linux-2.6.24/debian/rules.d/lpia.mk	2008-03-19 17:56:17.000000000 -0700
+++ /tmp/tmp.yfVsG13186/b/linux-2.6.24-12.22~zumappa/debian/rules.d/lpia.mk	2008-03-20 15:25:47.000000000 -0700
@@ -12,7 +12,7 @@
 loader		= grub
 
 #custom_flavours = lpiacompat lpia
-custom_flavours = lpia lpiacompat
+custom_flavours = lpia lpiacompat zumastor
 
 #
 # Set this variable to 'true' in order to

