diff -ru linux-2.6.24.2/mm/backing-dev.c linux-2.6.24.2-less.throttle.failed/mm/backing-dev.c
--- linux-2.6.24.2/mm/backing-dev.c	2008-02-10 21:51:11.000000000 -0800
+++ linux-2.6.24.2-less.throttle.failed/mm/backing-dev.c	2008-02-13 11:25:57.000000000 -0800
@@ -83,6 +83,10 @@
 	DEFINE_WAIT(wait);
 	wait_queue_head_t *wqh = &congestion_wqh[rw];
 
+	if ((current->flags & PF_LESS_THROTTLE) && printk_ratelimit()) {
+		printk(KERN_WARNING "PF_LESS_THROTTLE process %s[%i] entered congestion_wait!\n", current->comm, current->pid);
+		dump_stack();
+	}
 	prepare_to_wait(wqh, &wait, TASK_UNINTERRUPTIBLE);
 	ret = io_schedule_timeout(timeout);
 	finish_wait(wqh, &wait);
Only in linux-2.6.24.2-less.throttle.failed/mm: backing-dev.c.orig
