--- /var/tmp/zumastor-kernel-build/linux-2.6.24/debian/control	2008-03-17 14:11:09.000000000 -0700
+++ /var/tmp/build/new-pkg/linux-2.6.24/debian/control	2008-03-17 12:52:25.000000000 -0700
@@ -798,3 +798,25 @@
  .
  This is for sites that want the latest kernel headers.  Please read
  /usr/share/doc/linux-headers-2.6.24-12/debian.README.gz for details.
+
+Package: linux-image-2.6.24-12-zumastor
+Architecture: i386 amd64
+Section: universe/base
+Priority: optional
+Pre-Depends: dpkg (>= 1.10.24)
+Provides: linux-image, linux-image-2.6, fuse-module, kvm-api-4, redhat-cluster-modules
+Depends: initramfs-tools (>= 0.36ubuntu6), coreutils | fileutils (>= 4.0), module-init-tools (>= 3.3-pre11-4ubuntu3)
+Conflicts: hotplug (<< 0.0.20040105-1)
+Recommends: grub
+Suggests: fdutils, linux-doc-2.6.24 | linux-source-2.6.24
+Description: Linux kernel image for version 2.6.24 on This kernel can be used
+ for Zumastor hosts.
+
+Package: linux-headers-2.6.24-12-zumastor
+Architecture: i386 amd64
+Section: devel
+Priority: optional
+Depends: coreutils | fileutils (>= 4.0), linux-headers-2.6.24-12,  ${shlibs:Depends}
+Provides: linux-headers, linux-headers-2.6
+Description: Linux kernel headers for version 2.6.24 on These headers can be used
+ for Zumastor hosts.
